reframe/frontend/statistics.py

```diff
@@ -1,89 +1,103 @@
 
 
 class TestStats:
     """Stores test case statistics."""
 
-    def __init__(self, tasks=[]):
-        if not hasattr(tasks, '__iter__'):
-            raise TypeError('expected an iterable')
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = [[]]
+        self._current_run = 0
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
+    def next_run(self, current_run):
+        if current_run != self._current_run + 1:
+            raise ReframeFatalError('current_run variable out of sync')
+
+        self._current_run = current_run
+        self._tasks.append([])
+
+    def add_task(self, task):
+        self._tasks[self._current_run].append(task)
+
+    def _get_tasks(self, run):
+        try:
+            return self._tasks[run]
+        except IndexError:
+            raise StatisticsError('no such run: %s' % run)
+
+    def num_failures(self, run=-1):
+        return len([t for t in self._get_tasks(run) if t.failed])
+
+    def num_cases(self, run=-1):
+        return len(self._get_tasks(run))
+
+    def tasks_failed(self, run=-1):
+        return [t for t in self._get_tasks(run) if t.failed]
+
+    def retry_report(self):
+        # Return an empty report if no retries were done.
+        if not self._current_run:
+            return ''
+
+        line_width = 78
+        report = [line_width * '=']
+        report.append('SUMMARY OF RETRIES')
+        report.append(line_width * '-')
+        messages = {}
+        for run in range(1, len(self._tasks)):
+            for t in self._get_tasks(run):
+                # Overwrite entry from previous run if available
+                messages[t.check.info()] = (
```

Comment:
You should not rely on `check.info()` to hash the messages! This may be changed by the user.

Reply:
You should manually create a key out of the check name and the partition name as follows:

```python
key = '%s:%s:%s' % (t.check.name, t.check.current_partition.fullname, t.check.current_environ.name)
```

Reply:
Then, I should also use these explicit properties instead of `check.info()` to create the message?

Reply:
No, for the message to print you don't need to care. I've made the comment because you were hashing the message, which may not be unique.

Reply:
Done.